Update App Types document

thesmithee opened the pull request:
As agreed with Ron, updating this document to point to the deployment document. Also added a sub-section in the self-contained applications section to talk a bit about the smaller application creation and again point to Ron's document that will be describing that.
Fixes #Issue_Number

@rpetrusha @richlander @mairaw @BillWagner

Commit:
Update App Types document

As agreed with Ron, updating this document to point to the deployment document. Also added a sub-section in the self-contained applications section to talk a bit about the smaller application creation and again point to Ron's document that will be describing that.

svick commented:
> Also added a sub-section in the self-contained applications section to talk a bit about the smaller application creation and again point to Ron's document that **will** be describing that.

Shouldn't that link be added only *after* the content it links to actually exists? And then link to a specific section of that article?

blackdwarf commented:
Or, we can update and merge this PR after Ron finishes the deployment section and adds it. Seems to me both are good options. :)

Thanks,
Z

Sent from my tricorder




On Tue, Jul 19, 2016 at 5:45 PM -0700, "Petr Onderka" <notifications@github.com<mailto:notifications@github.com>> wrote:


Also added a sub-section in the self-contained applications section to talk a bit about the smaller application creation and again point to Ron's document that will be describing that.

Shouldn't that link be added only after the content it links to actually exists? And then link to a specific section of that article?

-
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub<https://github.com/dotnet/core-docs/pull/823#issuecomment-233808345>, or mute the thread<https://github.com/notifications/unsubscribe-auth/AC-H_Dux4LUUDK_y2Ps8-AcDvNBZB25Mks5qXW_CgaJpZM4JQTVk>.

rpetrusha commented on docs/core/app-types.md:
```diff
@@ -64,6 +63,9 @@ The `Microsoft.NETCore.App` is a "metapackage" that states that you are targetin
 property on that dependency means that at publish time, the tooling will skip publishing the assemblies for that dependency 
 to the published output. You don't need these since they will be installed with .NET Core on the targeted machine. 
 
+For an example and a how-to article on how to deploy applications please consult the [deploying applications](deploying/index.md) 
```
add a comma after "applications"

richlander replied:
Would be better to use active voice.

> See [Deploying applications](deploying/index.md) to learn how to deploy .NET Core applications.

rpetrusha commented on docs/core/app-types.md:
````diff
@@ -138,3 +140,10 @@ The following `project.json` sample illustrates a simple self-contained applicat
 ```
 
+### Making a smaller self-contained application
+As mentioned, self-contained application will package the entire runtime with the application. The "entire runtime" here 
````
better: "a self-contained application packages..."

richlander replied:
This section doesn't seem super useful. It describes the problem and then sends you somewhere else. Feels like we might have a factoring problem between the two documents.

rpetrusha replied:
I think that the duplication is acceptable. Given that app type and deployment are tightly coupled, you have to decide up front what kind of app you're going to create. 
The [deployment topic](https://docs.microsoft.com/en-us/dotnet/articles/core/deploying/index) covers both portable and self-contained apps. It could be broken into two tutorials, but generally I find it more helpful to have longer topics that can be used both to provide a comprehensive overview as well as step-by-step instructions for a particular deployment type.

rpetrusha commented:
With two corrections, LGTM.

richlander commented on docs/core/app-types.md:
```diff
@@ -34,8 +34,7 @@ There are two main types that we can observe:
 
 ## Portable applications
```
I'm thinking we should find a different term than "portable" since these apps are not portable in the industry sense.

rpetrusha replied:
What would be a better term to describe them? "Portable" here is at least descriptive -- the apps are portable from one .NET Core installation to another, given that they depend on the external runtime and class libraries. I can't think of a better term. Should we open an issue?

richlander commented:
Would love to rename the file from "app-types" to something more about publishing or deployment.

BillWagner commented:
@blackdwarf Is this ready to merge, or will there be further changes?

blackdwarf commented:
@BillWagner further changes as agreed with @rpetrusha and @richlander. I just need to find the time to sit down and write. 😃 🎉

rpetrusha commented:
Hi Z,

Since I just finished most of the work for the Framework 4.6.2 release, I should have some time to work on this tomorrow and Friday, if you’re too busy. (I’m on vacation next week.)

--Ron


From: Zlatko Knezevic [mailto:notifications@github.com]
Sent: Tuesday, August 02, 2016 12:16 PM
To: dotnet/core-docs
Cc: Ron Petrusha; Mention
Subject: Re: [dotnet/core-docs] Update App Types document (#823)


@BillWagner<https://github.com/BillWagner> further changes as agreed with @rpetrusha<https://github.com/rpetrusha> and @richlander<https://github.com/richlander>. I just need to find the time to sit down and write. 😃 🎉

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/dotnet/core-docs/pull/823#issuecomment-237012703>, or mute the thread<https://github.com/notifications/unsubscribe-auth/AKYfMSJM2vwzcBS7DW3US1Uc1O1pr2Fwks5qb5dWgaJpZM4JQTVk>.

mairaw commented:
where we are with this PR @blackdwarf? Is this WIP or should I do a review and see if it's ready to be merged?

blackdwarf commented:
@mairaw will be working on this this week. It is WIP.

Commit:
Finalizing the application types change

The following changes have been made:
* The app-types document is unlinked from the TOC
* The app-types document contains only a link to the deployment topic
 (since the document is linked across the various issues on GH and so on)
* Some content was folded into the deployment topic
* Changed from using the "application types" name to using "deployment
 types": framework-dependent deployment (ex-portable) and self-contained
 deployment

blackdwarf commented:
@mairaw @rpetrusha @richlander @BillWagner this is now ready for review.

qinezh commented:
Open Publishing Build Service: The pull request content has been published and here are some sample preview links:
- Docset 1: [core-docs](https://stage.docs.microsoft.com/en-us/dotnet/core/api/system.security.accesscontrol.directorysecurity?branch=pr-en-us-823)

BillWagner commented:
closing and re-opening to kick off the build again.

BillWagner commented on docs/core/app-types.md:
````diff
@@ -15,126 +15,5 @@ ms.assetid: 93488a0b-c94c-4ed6-97ea-571bb23a862e
-```
-
+## Content moved!
+This content has merged and moved to the [.NET Core Application Deployment](deployment/index.md) topic. 
````
I think this link should be (deploying/index.md)

BillWagner commented on docs/core/deploying/index.md:
```diff
@@ -14,21 +14,19 @@ ms.assetid: da7a31a0-8072-4f23-82aa-8a19184cb701
+- Framework-dependent deployment. As the name implies, framework-dependent deployment (FDD for short) relies on a shared system-wide version of .NET Core to be present on the target system. Because .NET Core is already present, your app is also portable between installations of .NET Core. Your app contains only its own code and any third-party dependencies that are outside of the .NET Core librar ...
 
-- As a self-contained application. Unlike a portable app, a self-contained app does not rely on any shared components to be present on the target system. All components, including both .NET Core libraries and the .NET Core runtime, are included with the application and are isolated from other .NET Core applications. Self-contained applications include an executable (such as `app.exe` on Windows p ...
+- Self-contained deployment. Unlike FDD, a self-contained deployment (SCD for short) does not rely on any shared components to be present on the target system. All components, including both .NET Core libraries and the .NET Core runtime, are included with the application and are isolated from other .NET Core applications. SCDs include an executable (such as `app.exe` on Windows platforms for a Se ...
```
typo: deploymentlication

rpetrusha replied:
remove extra space between "is" and "a"

BillWagner commented:
Overall, LGTM.
Found a couple small edits (1 bad link, 1 typo).
:shipit: once those are addressed.

(Removed the WIP on the title of the PR above).

qinezh commented:
Open Publishing Build Service: The pull request content has been published and here are some sample preview links:
- Docset 1: [core-docs](https://stage.docs.microsoft.com/en-us/dotnet/core/api/system.componentmodel.browsableattribute?branch=pr-en-us-823)

rpetrusha commented on docs/core/app-types.md:
```diff
@@ -15,126 +15,5 @@ ms.assetid: 93488a0b-c94c-4ed6-97ea-571bb23a862e
-a general purpose development platform.
-
-However, given its unique and cross-platform nature, .NET Core also has another angle through which to observe the type of the application 
-and that is the application's *portability*. Portability essentially means where you can run your application and what 
```
better: ...to observe the type of the application: the application's *portability*.

rpetrusha replied:
I'd change "means" to "defines"

mairaw replied:
@rpetrusha this is the old content

rpetrusha commented on docs/core/app-types.md:
```diff
@@ -15,126 +15,5 @@ ms.assetid: 93488a0b-c94c-4ed6-97ea-571bb23a862e
-However, given its unique and cross-platform nature, .NET Core also has another angle through which to observe the type of the application 
-and that is the application's *portability*. Portability essentially means where you can run your application and what 
-prerequisites you need to satisfy in order for your application to run on a given machine.
-This document deals with this angle, portability, and outlines the two main types of portability that .NET Core enables. 
```
This documents deals with portability and outlines...

rpetrusha commented on docs/core/app-types.md:
```diff
@@ -15,126 +15,5 @@ ms.assetid: 93488a0b-c94c-4ed6-97ea-571bb23a862e
-A subgroup of the above, this type is a portable application that has native dependencies specified 
-somewhere in its dependency chain. This application is as portable as all of its native dependencies 
-are portable. You will be able to run the application on any platform that your native dependencies can 
-run on. Prime example of this is Kestrel, the ASP.NET Core cross-platform web server. It is built on top of 
```
The prime

rpetrusha commented on docs/core/app-types.md:
```diff
@@ -15,126 +15,5 @@ ms.assetid: 93488a0b-c94c-4ed6-97ea-571bb23a862e
-somewhere in its dependency chain. This application is as portable as all of its native dependencies 
-are portable. You will be able to run the application on any platform that your native dependencies can 
-run on. Prime example of this is Kestrel, the ASP.NET Core cross-platform web server. It is built on top of 
-[libuv](https://github.com/libuv/libuv) which is its native dependency. 
```
), which is...

rpetrusha commented on docs/core/app-types.md:
```diff
@@ -15,126 +15,5 @@ ms.assetid: 93488a0b-c94c-4ed6-97ea-571bb23a862e
-the published output will contain a folder for each [Runtime Identifier (RID)](rid-catalog.md#what-are-rids) that the native dependency supports 
-(and that exists in its NuGet package). 
-
-The below `project.json` sample is showing an example of a portable application with a native dependency. 
```
The `project.json` sample below shows an...

rpetrusha commented on docs/core/app-types.md:
```diff
@@ -15,126 +15,5 @@ ms.assetid: 93488a0b-c94c-4ed6-97ea-571bb23a862e
-## Self-contained application
-Unlike the portable application, a *self-contained application* does not rely on any shared component to 
-be present on the machine where you want to deploy the application. As its name implies, it means that 
-the entire dependency closure, **including the runtime** is packaged with the application. This makes 
```
comma after runtime

rpetrusha commented on docs/core/app-types.md:
```diff
@@ -15,126 +15,5 @@ ms.assetid: 93488a0b-c94c-4ed6-97ea-571bb23a862e
-native dependencies, whether it has .NET Core installed or not. This makes it that much 
-easier to deploy to the target machine, since you only deploy your application. 
-
-Since the application carries the runtime within itself, you need to make an explicit choice which platforms your application 
```
instead of "to make an explicit choice which": "to explicitly choose which"

rpetrusha commented on docs/core/app-types.md:
```diff
@@ -15,126 +15,5 @@ ms.assetid: 93488a0b-c94c-4ed6-97ea-571bb23a862e
-not work on macOS or Linux and vice versa. Of course, you can add or remove platforms during development at any given time. 
-
-There are several steps to get to a self-contained application. The first is to remove any `"type": "platform"` properties 
-off of any dependencies you have. Second is to leave the dependency on `Microsoft.NETCore.App` as it will pull in
```
The second

rpetrusha replied:
The second is... Also, `Microsoft.NETCore.App`, since it

rpetrusha commented on docs/core/app-types.md:
```diff
@@ -15,126 +15,5 @@ ms.assetid: 93488a0b-c94c-4ed6-97ea-571bb23a862e
-
-There are several steps to get to a self-contained application. The first is to remove any `"type": "platform"` properties 
-off of any dependencies you have. Second is to leave the dependency on `Microsoft.NETCore.App` as it will pull in
-all of the rest of things that are needed. 
```
the things

rpetrusha commented on docs/core/app-types.md:
```diff
@@ -15,126 +15,5 @@ ms.assetid: 93488a0b-c94c-4ed6-97ea-571bb23a862e
-off of any dependencies you have. Second is to leave the dependency on `Microsoft.NETCore.App` as it will pull in
-all of the rest of things that are needed. 
-
-Finally, you need to add a `runtimes` node in your `project.json` that will list out the 
```
"lists" instead of "will list out"

rpetrusha commented on docs/core/app-types.md:
```diff
@@ -15,126 +15,5 @@ ms.assetid: 93488a0b-c94c-4ed6-97ea-571bb23a862e
-
-Finally, you need to add a `runtimes` node in your `project.json` that will list out the 
-[RIDs](rid-catalog.md#what-are-rids) you wish to use. When restoring a project that has the `runtimes` node in it, NuGet 
-will restore the needed runtime for all the RIDs specifies. Then, when you want to publish your application for a given platform,
```
specified, not specifies

rpetrusha commented on docs/core/app-types.md:
```diff
@@ -15,126 +15,5 @@ ms.assetid: 93488a0b-c94c-4ed6-97ea-571bb23a862e
-
-If you want to publish for the RID that represents the operating system you are using the [.NET Core SDK](sdk.md) 
-on, you don't have to specify anything to `dotnet publish`. However, you still have to specify that RID in your 
-`project.json` in order to get a standalone application. 
```
a standalone application for that platform.

rpetrusha commented on docs/core/app-types.md:
```diff
@@ -15,126 +15,5 @@ ms.assetid: 93488a0b-c94c-4ed6-97ea-571bb23a862e
-on, you don't have to specify anything to `dotnet publish`. However, you still have to specify that RID in your 
-`project.json` in order to get a standalone application. 
-
-> An important thing to note is that in the .NET Core Tools Preview 2 timeframe, the self-contained application is being published 
```
I wonder if this can be rephrased -- I'm not sure how many developers will realize that the tools and library versioning are not in sync, or really know what .NET Core Tools Preview 2 means. And this will have to be revised when tooling changes anyway, so maybe it's ok to just say currently, the self-contained application is published from...

rpetrusha commented on docs/core/app-types.md:
```diff
@@ -15,126 +15,5 @@ ms.assetid: 93488a0b-c94c-4ed6-97ea-571bb23a862e
-
-> An important thing to note is that in the .NET Core Tools Preview 2 timeframe, the self-contained application is being published 
-> from the NuGet cache on your machine. This means that all dependencies, including the actual .NET Core runtime and 
-> libraries, is not ready-to-run optimized, which means that it will have lower overall performance than portable 
```
...optimized and will have... (Avoid the repetition of "<something> means that".)

rpetrusha commented on docs/core/deploying/index.md:
```diff
@@ -119,17 +117,15 @@ Deploying a portable app with one or more third-party dependencies involves thre
+## Self-contained deployments (SCD) ##
 
-For a self-contained app, you deploy not only your app and any third-party dependencies, but the version of .NET Core that you build your app with.
+For a self-contained deployment, you deploy not only your app and any third-party dependencies, but the version of .NET Core that you build your app with. Creating an SCD does not, however, include the native dependencies of .NET Core itself on various platforms (for example, OpenSSL on macOS) so these need to be installed before running the application. 
```
Is there a list of native dependencies of .NET Core that can be cross-referenced here?

blackdwarf replied:
Good point @rpetrusha fixed.

rpetrusha commented on docs/core/deploying/index.md:
```diff
@@ -139,13 +135,15 @@ It also has a number of disadvantages:
 
 - Because .NET Core is included in your deployment package, you must select the target platforms for which you build deployment packages in advance.
 
+- Presently, the .NET Core runtime files are published from the NuGet cache, which means that they are not crossgen-ed which can impact performance of your application. 
```
crossgen-ed. This can impact the performance...

blackdwarf replied:
Fixed.

rpetrusha commented on docs/core/deploying/index.md:
```diff
@@ -316,13 +302,13 @@ The following is the complete project.json file for this project:
 When you deploy your application, any third-party dependencies used in your app are also contained with your application files. Third-party libraries do not already have to be present on the system on which the app is running.
 
-Note that you can only deploy a self-contained app with a third-party library to platforms supported by that library.
+Note that you can only deploy a self-contained deployment with a third-party library to platforms supported by that library. This is similar to having third-party dependencies with native dependencies in your framework-dependenent deployment. 
```
typo: dependent

rpetrusha commented on docs/core/deploying/index.md:
```diff
@@ -406,4 +392,4 @@ The following is the complete `project.json` file for this project.
 ## <a name="crossgen"></a> Native image Generation ##
 
-NET Core uses a just in time (JIT) compiler that stores application code in an intermediate format and compiles it to native code at runtime.  To increase startup performance, the shared framework is pre-compiled using a tool called `crossgen`.  To improve performance of your application, you can use the same tool on your application's binaries.  Its performance impact is more noticeable when dep ...
+NET Core uses a just in time (JIT) compiler that stores application code in an intermediate format and compiles it to native code at runtime.  To increase startup performance, the shared framework is pre-compiled using a tool called `crossgen`.  To improve performance of your application, you can use the same tool on your application's binaries.  Its performance impact is more noticeable when dep ...
```
typo: deployment instead of deploymentlication (although I like the word)

Commit:
Responding to PR feedback

qinezh commented:
Open Publishing Build Service: The pull request content has been published and here are some sample preview links:
- Docset 1: [core-docs](https://stage.docs.microsoft.com/en-us/dotnet/core/api/system.servicemodel.channels.bindingelementcollection?branch=pr-en-us-823)

mairaw commented on docs/core/deploying/index.md:
```diff
@@ -14,21 +14,19 @@ ms.assetid: da7a31a0-8072-4f23-82aa-8a19184cb701
+You can create two types of deployments for .NET Core applications: 
 
-- As a portable app. A portable app relies on a shared system-wide version of .NET Core to be present on the target system. Because .NET Core is already present, your app is portable between installations of .NET Core. Your app contains only its own code and any third-party dependencies that are outside of the .NET Core libraries. Portable applications are .dll files that can be launched by using ...
+- Framework-dependent deployment. As the name implies, framework-dependent deployment (FDD for short) relies on a shared system-wide version of .NET Core to be present on the target system. Because .NET Core is already present, your app is also portable between installations of .NET Core. Your app contains only its own code and any third-party dependencies that are outside of the .NET Core librar ...
```
acronyms are introduced without the need to say for short:
framework-dependent deployment (FDD for short) -> framework-dependent deployment (FDD)

blackdwarf replied:
Fixed.

mairaw commented on docs/core/deploying/index.md:
```diff
@@ -97,13 +95,13 @@ Deploying a portable app with no third-party dependencies simply involves buildi
 The complete set of application  files can be deployed in any way you'd like. For example, you can package them in a zip file, use a simple `copy` command, or deploy them with any installation package of your choice.
 
-Before deploying your app, you can also use `crossgen` to convert it to native code. However, its performance impact is smaller than for self-contained apps. For more information, see the [Native Image Generation](#crossgen) section.
+Before deploying your app, you can also use `crossgen` to convert it to native code. However, its performance impact is smaller than for Self-contained deployments. For more information, see the [Native Image Generation](#crossgen) section.
```
I think self should be lower caps

blackdwarf replied:
Fixed.

mairaw commented on docs/core/deploying/index.md:
```diff
@@ -14,21 +14,19 @@ ms.assetid: da7a31a0-8072-4f23-82aa-8a19184cb701
 
```
Update the ms.date to today's date to update the freshness on the topic

blackdwarf replied:
Done

mairaw commented on docs/core/deploying/index.md:
```diff
@@ -119,17 +117,15 @@ Deploying a portable app with one or more third-party dependencies involves thre
-Note that a portable app with third-party dependencies will only be as portable as its third-party dependencies. For example, if a third-party library only supports macOS, the app will not be portable to Windows systems.
-
-When you deploy your application, any third-party dependencies used in your app are resolved from the local NuGet cache of the system on which your app is running. In other words, the third-party library must be present on the target machine for your app to run successfully.
+Note that a framework-dependent deployment with third-party dependencies will only be as portable as its third-party dependencies. For example, if a third-party library only supports macOS, the app will not be portable to Windows systems. This can happen if the third-party dependency itself depends on native code. A good example of this is Kestrel server. When an FDD is created for an application ...
```
link to RID topic should probably be ../rid-catalog.md#what-are-rids

Commit:
Responding to PR feedback

mairaw commented on docs/core/deploying/index.md:
```diff
@@ -316,13 +302,13 @@ The following is the complete project.json file for this project:
 If the availability of adequate storage space on target systems is likely to be an issue, you can reduce the overall footprint of your app by excluding some system components. To do this, you explicitly define the .NET Core components that your app includes in your project.json file.
 
-To create a self-contained app with a smaller footprint, start by following the first two steps for creating a self-contained app. Once you've run the `dotnet new` command and added the C# source code to your app, do the following:
+To create a Self-contained deployment with a smaller footprint, start by following the first two steps for creating a self-contained deployment. Once you've run the `dotnet new` command and added the C# source code to your app, do the following:
```
lowercase self

mairaw commented on docs/core/deploying/index.md:
```diff
@@ -346,9 +332,9 @@ This does two things:
+   This defines the system components used by our app. The system components packaged with our app include the .NET Standard Library, the .NET Core runtime, and the .NET Core host. This produces a self-contained deployment with a smaller footprint.
 
-3. As you did in the [Deploying a Simple Self-Contained App](#simpleSelf) example, create a `runtimes` section in your `project.json` file that defines the platforms your app targets and specify the runtime identifier of each platform that you target. See [Runtime IDentifier catalog](../rid-catalog.md) for a list of runtime identifiers. For example, the following `runtimes` section indicates that ...
+3. As you did in the [Deploying a Simple self-contained deployment](#simpleSelf) example, create a `runtimes` section in your `project.json` file that defines the platforms your app targets and specify the runtime identifier of each platform that you target. See [Runtime IDentifier catalog](../rid-catalog.md) for a list of runtime identifiers. For example, the following `runtimes` section indicat ...
```
lowercase simple

Commit:
Responding to PR feedback

qinezh commented:
Open Publishing Build Service: The pull request content has been published and here are some sample preview links:
- Docset 1: [core-docs](https://stage.docs.microsoft.com/en-us/dotnet/core/api/system.security.cryptography.md5?branch=pr-en-us-823)

qinezh commented:
Open Publishing Build Service: The pull request content has been published and here are some sample preview links:
- Docset 1: [core-docs](https://stage.docs.microsoft.com/en-us/dotnet/core/api/system.security.cryptography.cngpropertycollection?branch=pr-en-us-823)

blackdwarf commented:
OK, seems that stuff is done. We ready to merge this?

rpetrusha commented:
Yes.

mairaw commented:
Thanks @blackdwarf!
